wasm/validation/context.py

```diff
@@ -35,7 +36,110 @@
+    tables: Tuple[TableType, ...]
+    mems: Tuple[MemoryType, ...]
+    globals: Tuple[GlobalType, ...]
+    locals: Tuple[ValType, ...]
+    labels: Tuple[ValType, ...]
+    # `returns` may be `None`, but the argument must be
+    # explicitely supplied.
+    returns: Union[None, Tuple[ValType, ...]]
+
+    #
+    # Types
+    #
+    def validate_type_idx(self, idx: TypeIdx) -> None:
+        if not self.has_type(idx):
+            raise ValidationError(
+                f"Types index outside of valid range: {idx} >= {len(self.types)}"
+            )
+
+    def has_type(self, idx: TypeIdx) -> bool:
+        return idx < len(self.types)
+
+    def get_type(self, idx: TypeIdx) -> FunctionType:
+        return self.types[idx]
+
+    #
+    # Functions
+    #
+    def validate_function_idx(self, idx: FuncIdx) -> None:
+        if not self.has_function(idx):
+            raise ValidationError(
+                f"Function index outside of valid range: {idx} >= {len(self.functions)}"
+            )
+
+    def has_function(self, idx: FuncIdx) -> bool:
+        return idx < len(self.functions)
+
+    def get_function(self, idx: FuncIdx) -> FunctionType:
+        return self.functions[idx]
+
+    #
+    # Tables
+    #
+    def validate_table_idx(self, idx: TableIdx) -> None:
+        if not self.has_table(idx):
+            raise ValidationError(
+                f"Table index outside of valid range: {idx} >= {len(self.tables)}"
+            )
+
+    def has_table(self, idx: TableIdx) -> bool:
+        return idx < len(self.tables)
+
+    def get_table(self, idx: TableIdx) -> TableType:
+        return self.tables[idx]
+
+    #
+    # Memory
+    #
+    def validate_mem_idx(self, idx: MemoryIdx) -> None:
+        if not self.has_mem(idx):
+            raise ValidationError(
+                f"Memory index outside of valid range: {idx} >= {len(self.mems)}"
+            )
+
+    def has_mem(self, idx: MemoryIdx) -> bool:
+        return idx < len(self.mems)
+
+    def get_mem(self, idx: MemoryIdx) -> MemoryType:
+        return self.mems[idx]
+
+    #
+    # Global
+    #
+    def validate_global_idx(self, idx: GlobalIdx) -> None:
+        if not self.has_global(idx):
+            raise ValidationError(
+                f"Global index outside of valid range: {idx} >= {len(self.globals)}"
+            )
+
+    def has_global(self, idx: GlobalIdx) -> bool:
+        return idx < len(self.globals)
+
+    def get_global(self, idx: GlobalIdx) -> GlobalType:
+        return self.globals[idx]
+
+    #
+    # Locals
+    #
+    def validate_local_idx(self, idx: LocalIdx) -> None:
+        if not self.has_local(idx):
+            raise ValidationError(
+                f"Locals index outside of valid range: {idx} >= {len(self.locals)}"
+            )
+
+    def has_local(self, idx: LocalIdx) -> bool:
+        return idx < len(self.locals)
+
+    def get_local(self, idx: LocalIdx) -> ValType:
+        return self.locals[idx]
+
+
+class Context(BaseContext):
```

Comment:
Alternative name brainstorm:
- `Inventory`
- ~`Register`~
- ~`Record`~
- `Lookup`
- ~`Index`~
- ~`Tabulation`~
- `Catalog`

I'm focusing on the idea that it's sort of a complete register of all items, and a mechanism to look them up. A few are struck out for being too heavily used in other places, or confused for verbs.